python/hail/methods/statgen.py

```diff
@@ -1100,3 +1100,192 @@ def rrm(call_expr):
                                             [row.s for row in dataset.cols_table().select('s').collect()],
                                             n_variants)
+
+class FilterAlleles(object):
+    """Filter out a set of alternate alleles.  If all alternate alleles of
+    a variant are filtered out, the variant itself is filtered out.
+    `filter_expr` is an alternate allele indexed `Array[Boolean]`
+    where the booleans, combined with `keep`, determine which
+    alternate alleles to filter out.
+
+    This object has bindings for values used in the filter alleles
+    process: `old_to_new`, `new_to_old` and `new_v`.  Call
+    :meth:`.FilterAlleles.annotate_rows` and/or
+    :meth:`.FilterAlleles.annotate_entries` to update row-indexed and
+    row- and column-indexed fields for the new, allele filtered
+    variant.  Finally, call :meth:`.FilterAlleles.filter` to perform
+    filter alleles.
+
+    Examples
+    --------
+    
+    Filter alleles with zero AC count on a dataset with the HTS entry
+    schema and update the ``info.AC`` and entry fields.
+
+    >>> fa = methods.FilterAlleles(dataset.info.AC.map(lambda AC: AC == 0), keep=False)
+    ... fa.annotate_rows(
+    ...     info = dataset.info.annotate(AC = fa.new_to_old[1:].map(lambda i: dataset.info.AC[i - 1])))
+    ... newPL = functions.cond(
+    ...     functions.is_defined(dataset.PL),
+    ...     functions.range(0, fa.new_v.num_genotypes()).map(
+    ...         lambda newi: functions.bind(
+    ...             functions.call(newi),
+    ...             lambda newc: dataset.PL[functions.gt_index(fa.new_to_old[newc.gtj()], fa.new_to_old[newc.gtk()])])),
+    ...     functions.null(TArray(TInt32())))
+    ... fa.annotate_entries(
+    ...     GT = functions.call(functions.gt_from_pl(newPL)),
+    ...     AD = functions.cond(
+    ...         functions.is_defined(dataset.AD),
+    ...         functions.range(0, fa.new_v.num_alleles()).map(
+    ...             lambda newi: dataset.AD[fa.new_to_old[newi]]),
+    ...         functions.null(TArray(TInt32()))),
+    ...     GQ = functions.gq_from_pl(newPL),
+    ...     PL = newPL)
+    ... filtered_result = fa.filter()
+    
+    Parameters
+    ----------
+    filter_expr : :class:`.ArrayBooleanExpression`
+        Boolean filter expression.
+
+    keep : bool
+        If True, keep alternate alleles where the corresponding
+        element of `filter_expr` is True.  If False, remove the
+        alternate alleles where the corresponding element is True.
+
+    left_aligned : bool
+        If True, variants are assumed to be left aligned and have
+        unique loci.  This avoids a shuffle.  If the assumption is
+        violated, an error is generated.
+
+    keep_star : bool
+        If True, keep variants where the only unfiltered alternate
+        alleles are ``*`` alleles.
+
+    """
+    
+    @typecheck_method(filter_expr=ArrayBooleanExpression, keep=bool, left_aligned=bool, keep_star=bool)
```

Comment:
expr_list, check type below

Reply:
we'll add better expr checking later

Reply:
I don't understand this suggestion.  We have very nice generic type checking logic, why not use it?  You don't like saying `ArrayBooleanExpression` compared to "an expression of type `Array[Boolean]`"?